components/login/login_controller/login_controller.jsx

```diff
@@ -233,18 +236,35 @@ export default class LoginController extends React.Component {
     }
 
     finishSignin(team) {
+        let redirectAction;
+        let redirectURL;
         const experimentalPrimaryTeam = this.props.experimentalPrimaryTeam;
         const primaryTeam = TeamStore.getByName(experimentalPrimaryTeam);
         const query = new URLSearchParams(this.props.location.search);
         const redirectTo = query.get('redirect_to');
 
         GlobalActions.loadCurrentLocale();
         if (redirectTo && redirectTo.match(/^\/([^/]|$)/)) {
-            browserHistory.push(redirectTo);
+            redirectURL = decodeURIComponent(redirectTo);
         } else if (team) {
-            browserHistory.push(`/${team.name}`);
+            redirectURL = `/${team.name}`;
         } else if (primaryTeam) {
-            browserHistory.push(`/${primaryTeam.name}/channels/${Constants.DEFAULT_CHANNEL}`);
+            redirectURL = `/${primaryTeam.name}/channels/${Constants.DEFAULT_CHANNEL}`;
+        } else {
+            redirectAction = Constants.DEFAULT_TEAM_REDIRECT;
+        }
+
+        if (this.props.showTermsOfService) {
```

Comment:
Here, we could avoid writing this, but by writing this here, we do save one redirect as otherwise the user would be redirected to the default team and channel after login, before being redirected to '/terms_of_service' page. 

I don't know how or if that would impact performance or user experience. We can remove this and revert to the previous version if needed.

Reply:
Since agreeing to the terms of service is a rare event, I don't think performance will be an issue. 0/5 at this point :)

Reply:
Can we remove this code if it is not needed? The added complexity cost is not worth the marginal performance gain for a rare event.